### PR TITLE
test version early test event

### PR DESCRIPTION
**Pull request by kish1n**

### CODE REVIEW

File: internal/cli/event.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	"log"
+
+	"github.com/rarimo/geo-points-svc/internal/config"
+	"github.com/rarimo/geo-points-svc/internal/service/event"
+)
+
+func eventStart(cfg config.Config, data int) {
+	err := event.Run(cfg, data)
+	if err != nil {
+		log.Fatalf("Error starting event: %v", err)
```
Comment (Zaptoss) [inline internal/cli/event.go:13]: Use `cfg.Log().Error` instead `log.Fatalf`

File: internal/cli/main.go [hunk 2]
```diff
@@ -29,6 +29,10 @@ func Run(args []string) bool {
 		migrateCmd     = app.Command("migrate", "migrate command")
 		migrateUpCmd   = migrateCmd.Command("up", "migrate db up")
 		migrateDownCmd = migrateCmd.Command("down", "migrate db down")
+
+		event    = app.Command("event", "claim event command")
+		eventCmd = event.Command("start", "claim event command")
+		data     = eventCmd.Arg("data", "data after ...").Required().Int()
```
Comment (Zaptoss) [inline internal/cli/main.go:35]: Change command names to more understandable: `event create-early-test -before 1730000`

File: internal/data/balances.go [hunk 3]
```diff
@@ -54,6 +54,8 @@ type BalancesQ interface {
 	FilterByInternalAID(aid string) BalancesQ
 	FilterByExternalAID(aid string) BalancesQ
 	FilterBySharedHash(hash string) BalancesQ
+	FilterByCreatedBefore(time int) BalancesQ
```
Comment (Zaptoss) [inline internal/data/balances.go:57]: Use another name. time - standard go package

File: internal/data/pg/balances.go [hunk 4]
```diff
@@ -222,10 +222,18 @@ func (q *balances) FilterBySharedHash(hash string) data.BalancesQ {
...
 func (q *balances) applyCondition(cond squirrel.Sqlizer) data.BalancesQ {
 	q.selector = q.selector.Where(cond)
 	q.updater = q.updater.Where(cond)
 	q.rank = q.rank.Where(cond)
 	q.counter = q.counter.Where(cond)
 	return q
 }
+
+func (q *balances) FilterVerified() data.BalancesQ {
+	return q.applyCondition(squirrel.Gt{"verified": true})
```
Comment (Zaptoss) [inline internal/data/pg/balances.go:238]: balances table don't have `verified` field.

File: internal/service/event/main.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	if users == nil {
+		log.Infof("no users found")
+		return nil
+	}
+
+	for _, user := range users {
+		userNull := user.Nullifier
+		eve, err := filtEvents.FilterByNullifier(userNull).Get()
```
Comment (Zaptoss) [inline internal/service/event/main.go:34]: Calling the database for each user is very inefficient, you need to get all the events of this type at once and then go through them in a loop

**Comment by Zaptoss** [comment]
It is necessary to check whether there is such an event in the configuration and whether it is active. Also, after adding an event, you need to check if the auto_claim flag is set and claim these events for users

File: internal/cli/main.go [hunk 6]
```diff
@@ -29,6 +29,10 @@ func Run(args []string) bool {
 		migrateCmd     = app.Command("migrate", "migrate command")
 		migrateUpCmd   = migrateCmd.Command("up", "migrate db up")
 		migrateDownCmd = migrateCmd.Command("down", "migrate db down")
+
+		event    = app.Command("event", "claim event command")
+		eventCmd = event.Command("create-early-test", "claim event command")
+		data     = eventCmd.Arg("-before", "data after ...").Required().Int()
```
Comment (Zaptoss) [inline internal/cli/main.go:35]: Just `before`. In previous comment it was as a example, because when you pass parameters, you need use -paramname

File: internal/data/pg/balances.go [hunk 7]
```diff
@@ -222,10 +222,18 @@ func (q *balances) FilterBySharedHash(hash string) data.BalancesQ {
...
 func (q *balances) applyCondition(cond squirrel.Sqlizer) data.BalancesQ {
 	q.selector = q.selector.Where(cond)
 	q.updater = q.updater.Where(cond)
 	q.rank = q.rank.Where(cond)
 	q.counter = q.counter.Where(cond)
 	return q
 }
+
+func (q *balances) FilterVerified() data.BalancesQ {
+	return q.applyCondition(squirrel.Gt{"is_verified": true})
```
Comment (Zaptoss) [inline internal/data/pg/balances.go:238]: `balances` table don't have `is_verified` field. And you should use `suirrel.Eq`

File: internal/data/events.go [hunk 8]
```diff
@@ -65,4 +65,5 @@ type EventsQ interface {
 	FilterByUpdatedAtBefore(int64) EventsQ
 	FilterByExternalID(string) EventsQ
 	FilterInactiveNotClaimed(types ...string) EventsQ
+	FilterByNullifiers(nullifiers []string) EventsQ
```
Comment (Zaptoss) [inline internal/data/events.go:68]: You already have `FilterByNullifier`. You can modify existing function

File: internal/service/event/main.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+func Run(cfg config.Config, date int) error {
+	log := cfg.Log()
+	db := cfg.DB()
+
+	balances := pg.NewBalances(db)
+	events := pg.NewEvents(db)
+
+	users, err := balances.FilterByCreatedBefore(date).FilterVerified().Select()
```
Comment (Zaptoss) [inline internal/service/event/main.go:19]: This is `balances`, not `users`

File: internal/service/event/main.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to insert `early_test` event: %w", err)
+		}
+
+		var totalPoints int64
+
+		evType := evTypes.Get(models.TypeEarlyTest, evtypes.FilterInactive, evtypes.FilterByAutoClaim(true))
+		if evType == nil {
+			continue
+		}
```
Comment (Zaptoss) [inline internal/service/event/main.go:82]: You must check it once before loop

File: internal/service/event/main.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		log.WithError(err).Error("failed to filter by updated before")
+		return err
+	}
+	if len(balances) == 0 {
+		log.Infof("no balances found")
+		return nil
+	}
+
+	nullifiers := make([]string, 0)
```
Comment (Zaptoss) [inline internal/service/event/main.go:41]: `make([]string, 0, len(balances))`

File: internal/service/event/main.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	filteredEvents, err := eventsQ.
+		FilterByType(models.TypeEarlyTest).
+		FilterByStatus(data.EventFulfilled).
+		FilterByNullifier(nullifiers...).
+		Select()
+
+	if err != nil {
+		log.WithError(err).Error("failed to filter eventsQ")
```
Comment (Zaptoss) [inline internal/service/event/main.go:54]: Capitilize and use more correct message like "Failed to select %s events"

File: internal/service/event/main.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+	for _, balance := range balances {
+		nullifiers = append(nullifiers, balance.Nullifier)
+	}
+
+	filteredEvents, err := eventsQ.
+		FilterByType(models.TypeEarlyTest).
+		FilterByStatus(data.EventFulfilled).
+		FilterByNullifier(nullifiers...).
+		Select()
```
Comment (Zaptoss) [inline internal/service/event/main.go:51]: You shouldn't use filter by type, because user can have already claimed or opened events so you add second event for user.

File: internal/service/event/main.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	log := cfg.Log()
+	db := cfg.DB()
+	lvls := cfg.Levels()
+	evTypes := cfg.EventTypes()
+
+	balancesQ := pg.NewBalances(db)
+	eventsQ := pg.NewEvents(db)
+	referralsQ := pg.NewReferrals(db)
+
+	evType := evTypes.Get(models.TypeEarlyTest, evtypes.FilterInactive, evtypes.FilterByAutoClaim(true))
```
Comment (Zaptoss) [inline internal/service/event/main.go:24]: You shouldn't filter by autoclaim here, because if event active, we can add fulfilled event and if autoclaim=true - claim.

File: internal/service/event/main.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+		FilterByStatus(data.EventFulfilled).
+		FilterByNullifier(nullifiers...).
+		Select()
+
+	if err != nil {
+		log.WithError(err).Error("Failed to select %s events")
+		return err
+	}
+
+	EventsMap := make(map[string]data.Event)
```
Comment (Zaptoss) [inline internal/service/event/main.go:57]:
```suggestion
	eventsMap := make(map[string]string, len(filteredEvents))
```

File: internal/service/event/main.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+	for _, event := range filteredEvents {
+		EventsMap[event.Nullifier] = event
+	}
```
Comment (Zaptoss) [inline internal/service/event/main.go:60]:
```suggestion
	for _, event := range filteredEvents {
		eventsMap[event.Nullifier] = ""
	}
```

File: internal/service/event/main.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+			Type:      models.TypeEarlyTest,
+			Status:    data.EventFulfilled,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to insert `early_test` event: %w", err)
+		}
+
+		var totalPoints int64
+
+		_, err = eventsQ.FilterByNullifier(nullifiers...).Update(data.EventClaimed, nil, &evType.Reward)
```
Comment (Zaptoss) [inline internal/service/event/main.go:78]: Claim events only if autoclaim = true. nullifiers also contain balances which already have this event. Use `balance.Nullifier`

File: internal/service/event/main.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		err = balancesQ.FilterByNullifier(balance.Nullifier).Update(map[string]any{
+			data.ColAmount: pg.AddToValue(data.ColAmount, totalPoints),
+			data.ColLevel:  level,
+		})
+
+		if err != nil {
+			return fmt.Errorf("error update balance amount and level: %w", err)
+		}
```
Comment (Zaptoss) [inline internal/service/event/main.go:97]: Wrap this in transaction

File: internal/cli/main.go [hunk 19]
```diff
@@ -29,6 +29,10 @@ func Run(args []string) bool {
 		migrateCmd     = app.Command("migrate", "migrate command")
 		migrateUpCmd   = migrateCmd.Command("up", "migrate db up")
 		migrateDownCmd = migrateCmd.Command("down", "migrate db down")
+
+		event    = app.Command("event", "claim event command")
+		eventCmd = event.Command("create-early-test", "claim event command")
+		data     = eventCmd.Arg("before", "data after ...").Required().Int()
```
Comment (artemskriabin) [inline internal/cli/main.go:35]: date or data?
data after... what?

File: internal/service/event/main.go [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+		FilterByStatus(data.EventFulfilled).
+		FilterByNullifier(nullifiers...).
+		Select()
+
+	if err != nil {
+		log.WithError(err).Errorf("Failed to select %s events", err)
+		return err
+	}
+
+	eventsMap := make(map[string]string, len(filteredEvents))
```
Comment (artemskriabin) [inline internal/service/event/main.go:57]: if you need set use map[string]struct{}

File: internal/service/event/main.go [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+	nullifiers := make([]string, 0, len(balances))
+
+	for _, balance := range balances {
+		nullifiers = append(nullifiers, balance.Nullifier)
+	}
+
+	filteredEvents, err := eventsQ.
+		FilterByStatus(data.EventFulfilled).
+		FilterByNullifier(nullifiers...).
+		Select()
```
Comment (artemskriabin) [inline internal/service/event/main.go:50]: filter by type

File: internal/service/event/main.go [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+			if evtypes.FilterByAutoClaim(true)(*evType) {
+				return nil
+			}
+
+			var totalPoints int64
+
+			_, err = eventsQ.FilterByNullifier(balance.Nullifier).Update(data.EventClaimed, nil, &evType.Reward)
+			if err != nil {
+				return fmt.Errorf("failedt to update %s events for user=%s: %w", models.TypeEarlyTest, balance.Nullifier, err)
```
Comment (artemskriabin) [inline internal/service/event/main.go:85]: failed

File: internal/service/event/main.go [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+		return err
+	}
+
+	eventsMap := make(map[string]string, len(filteredEvents))
+	for _, event := range filteredEvents {
+		eventsMap[event.Nullifier] = ""
+	}
+
+	for _, balance := range balances {
+		err = eventsQ.Transaction(func() error {
```
Comment (artemskriabin) [inline internal/service/event/main.go:63]: call New() to clean filters

**Review by artemskriabin — CHANGES_REQUESTED** [review]

File: internal/cli/event.go [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+		go func() {
+			f()
+			wg.Done()
+		}()
+	}
+
+	run(func() { evtypes.InitForOneTimeEvent(cfg, evTypesSig) })
+	<-evTypesSig
+
+	fmt.Println("ini end")
```
Comment (artemskriabin) [inline internal/cli/event.go:26]: why not log?

File: internal/cli/event.go [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+	run := func(f func()) {
+		wg.Add(1)
+		go func() {
+			f()
+			wg.Done()
+		}()
+	}
+
+	run(func() { evtypes.InitForOneTimeEvent(cfg, evTypesSig) })
+	<-evTypesSig
```
Comment (artemskriabin) [inline internal/cli/event.go:24]: overcomplicated

File: internal/service/event/main.go [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+
+			if err != nil {
+				return fmt.Errorf("failed to insert `early_test` event: %w", err)
+			}
+
+			if evtypes.FilterByAutoClaim(true)(*evType) {
+				return nil
+			}
+
+			var totalPoints int64
```
Comment (artemskriabin) [inline internal/service/event/main.go:82]: why do you need this var?

File: internal/service/event/main.go [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+		Select()
+
+	if err != nil {
+		log.WithError(err).Errorf("Failed to select %s events", err)
+		return err
+	}
+
+	for _, balance := range balances {
+		err = eventsQ.New().Transaction(func() error {
+			for _, event := range filteredEvents {
```
Comment (artemskriabin) [inline internal/service/event/main.go:61]: why did you remove map?

File: internal/service/event/main.go (hunk 13 above)
Comment (artemskriabin) [reply internal/service/event/main.go:52]: @kish1n remove filter by status

**Review by artemskriabin — APPROVED** [review]
